### PR TITLE
Eth JSON-RPC: from in eth_getTransactionByHash is not correctly popul…

### PR DESCRIPTION
**Pull request by maciejwitowski**
…ated #1614

## Related Issues
Closes https://github.com/filecoin-project/ref-fvm/issues/1614

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: chain/types/ethtypes/eth_transactions.go [hunk 1]
```diff
@@ -46,6 +46,7 @@ type EthTxArgs struct {
 	ChainID              int         `json:"chainId"`
 	Nonce                int         `json:"nonce"`
 	To                   *EthAddress `json:"to"`
+	From                 *EthAddress `json:"from"`
```
Comment (arajasek) [inline chain/types/ethtypes/eth_transactions.go:49]: I don't think we want to add this to `EthTxArgs` --the transaction arguments are independent of any single sender -- it should be the case that whoever signs them has a valid tx ready to go.

File: api/api_full.go [hunk 2]
```diff
@@ -1345,7 +1345,7 @@ type EthTxReceipt struct {
 	TransactionIndex  ethtypes.EthUint64   `json:"transactionIndex"`
 	BlockHash         ethtypes.EthHash     `json:"blockHash"`
 	BlockNumber       ethtypes.EthUint64   `json:"blockNumber"`
-	From              ethtypes.EthAddress  `json:"from"`
+	From              *ethtypes.EthAddress `json:"from"`
```
Comment (arajasek) [inline api/api_full.go:1348]: What's motivating this change?

File: chain/types/ethtypes/eth_transactions.go [hunk 3]
```diff
@@ -122,6 +125,12 @@ func EthTxArgsFromUnsignedEthMessage(msg *types.Message) (EthTxArgs, error) {
 		}
 		to = &addr
 
+		fromAddr, err := EthAddressFromFilecoinAddress(msg.From)
```
Comment (arajasek) [inline chain/types/ethtypes/eth_transactions.go:128]: There's a slight issue with this -- we might not be returning the "best" possible EthAddress here. In particular, we'd like to resolve to _actual_ eth addresses if the `msg.From` is an ID address (by looking up the corresponding f4 address). See `node/impl/full/eth.go`'s `lookupEthAddress` for the best-effort implementation of this.

I think we want to make this change at the API level when forming an `EthTx`. I think `newEthTxFromSignedMessage` is likely the best place for this.

**Comment by arajasek** [comment]
Not sure what's wrong with CI, but it's definitely not caused by this PR :)

File: chain/types/ethtypes/eth_transactions.go [hunk 4]
```diff
@@ -77,10 +77,16 @@ func EthTxFromSignedEthMessage(smsg *types.SignedMessage) (EthTx, error) {
 		return EthTx{}, xerrors.Errorf("failed to recover signature: %w", err)
 	}
 
+	fromAddr, err := EthAddressFromFilecoinAddress(smsg.Message.From)
```
[the remark this answers is not in the document]
Comment (maciejwitowski) [reply chain/types/ethtypes/eth_transactions.go]: Pushing this 1 level up makes sense 👍

File: node/impl/full/eth.go [hunk 5]
```diff
@@ -1571,6 +1571,13 @@ func newEthTxFromSignedMessage(ctx context.Context, smsg *types.SignedMessage, s
 		if err != nil {
 			return ethtypes.EthTx{}, err
```
Comment (arajasek) [inline node/impl/full/eth.go:1572]: Nit: Can you please also include an error message here? It's _totally_ not related to this PR, but it's been frustrating not knowing where errors have been originating from in the Eth API (so I'm trying to slowly add them).
```suggestion
			return ethtypes.EthTx{}, xerrors.Errorf("failed to calculate hash for ethTx: %w", err)

```

**Review by arajasek — COMMENTED** [review]
Change looks good, but let's expand one of the tests to cover this case (clearly needed!)

File: node/impl/full/eth.go (hunk 5 above)
Comment (maciejwitowski) [reply node/impl/full/eth.go:1572]: Absolutely ❤️

**Review by arajasek — APPROVED** [review]
LGTM!

File: itests/eth_transactions_test.go [hunk 6]
```diff
@@ -83,6 +83,11 @@ func TestValueTransferValidSignature(t *testing.T) {
 
 	// Success.
 	require.EqualValues(t, ethtypes.EthUint64(0x1), receipt.Status)
+
+	ethTx, err := client.EthGetTransactionByHash(ctx, &hash)
+	require.Nil(t, err)
+	require.EqualValues(t, ethAddr, ethTx.From)
```
Comment (raulk) [inline itests/eth_transactions_test.go:89]: Let's also assert against all `tx` fields? That way if something else fails to populate/breaks, we'll catch it too.

**Review by raulk — APPROVED** [review]
Can strengthen the test.

File: itests/eth_transactions_test.go (hunk 6 above)
Comment (maciejwitowski) [reply itests/eth_transactions_test.go:89]: Won't be able to do it today, but happy to land this later this week 🙂

**Pull request merged**